vispy/scene/widgets/grid.py

```diff
@@ -183,63 +258,239 @@ def __repr__(self):
         return (('<Grid at %s:\n' % hex(id(self))) +
                 str(self.layout_array + 1) + '>')
 
-    def _update_child_widgets(self):
-        # Resize all widgets in this grid to share space.
-        n_rows, n_cols = self.grid_size
-        if n_rows == 0 or n_cols == 0:
-            return
+    @staticmethod
+    def _add_total_width_constraints(solver, width_grid, _var_w):
+        for (y, ws) in enumerate(width_grid):
+            width_expr = expression.Expression()
+            for (x, w) in enumerate(ws):
+                width_expr = width_expr + w
+            solver.add_constraint(width_expr == _var_w, strength=REQUIRED)
+
+    @staticmethod
+    def _add_total_height_constraints(solver, height_grid, _var_h):
+        for hs in height_grid:
+            height_expr = expression.Expression()
+            for h in hs:
+                height_expr += h
+            solver.add_constraint(height_expr == _var_h, strength=REQUIRED)
+
+    @staticmethod
+    def _add_gridding_width_constraints(solver, width_grid):
+        # access widths of one "y", different x
+        for ws in width_grid.T:
+            for w in ws[1:]:
+                solver.add_constraint(ws[0] == w, strength=REQUIRED)
+
+    @staticmethod
+    def _add_gridding_height_constraints(solver, height_grid):
+        # access heights of one "y"
+        for hs in height_grid.T:
+            for h in hs[1:]:
+                solver.add_constraint(hs[0] == h, strength=REQUIRED)
+
+    @staticmethod
+    def _add_stretch_constraints(solver, width_grid, height_grid,
+                                 grid_widgets, widget_grid):
+        xmax = len(height_grid)
+        ymax = len(width_grid)
+
+        stretch_widths = [[] for _ in range(0, ymax)]
+        stretch_heights = [[] for _ in range(0, xmax)]
+
+        for (_, val) in grid_widgets.items():
```

Comment:
I decided against destructuring it in the same line because it was way too much visual noise in the same line for my taste. I could change t if you want me to.